### PR TITLE
Remote improvements

### PR DESCRIPTION
**Pull request by apahim**
- Job: make test suite before the loader so the loader can know when it's a remoter execution to skip the discover.
- Remote: don't break on urls with test references.
- Invert the logics of `--remote-no-copy`, renaming it to `--remote-copy` and don't copy tests by default.

### CODE REVIEW

File: avocado/core/job.py [hunk 1]
```diff
@@ -285,6 +286,11 @@ def _make_test_suite(self, urls=None):
                      Optionally, a list of tests (each test a string).
         :returns: a test suite (a list of test factories)
         """
+        suite = [(None, {})]
```
Comment (clebergnu) [inline avocado/core/job.py]: This seems to work, and the code changes are simple and direct.

Still, I wonder if you pursued a different approach, more aimed at the loaders.  What I mean is, the entity responsible for resolving a collection of test names into a test suite is the loader.  In this example, we bypass the loader by creating a fake test suite (which shouldn't be the job direct responsibility).

I don't see a perfect way to do this without some other refactors, but, maybe `TestLoaderProxy.load_plugins()` could detect that a remote runner will be used, and thus, not load *any* plugin.  Then, `Job._make_test_suite()` should transparently return an empty suite.  The idea is to put the special logic where it's most appropriate, and all other used APIs should work be fulfilling their intended use.  When make runners and loaders proper plugins, the place where this special logic will be located should be much more clear.

Please let me know what you think of this suggestion.

File: avocado/core/remote/runner.py [hunk 2]
```diff
@@ -64,6 +64,8 @@ def _copy_files(self):
         paths = set()
         for i in xrange(len(self.job.urls)):
             url = self.job.urls[i]
+            if ':' in url:
```
Comment (clebergnu) [inline avocado/core/remote/runner.py]: Because of a typo in the commit message, I spent some time trying to find where/how this patch added a `_copy_files()` call/check/etc.  It should read: "This patch adds a check to _copy_files()...`

Then about the patch itself, it LGTM.  Let me do some functional testing.

File: avocado/core/remote/runner.py [hunk 3]
```diff
@@ -56,7 +56,7 @@ def _copy_files(self):
         $remote_test_dir + $test_absolute_path.
         :note: Default tests execution is translated into absolute paths too
         """
-        if self.job.args.remote_no_copy:    # Leave everything as is
+        if not self.job.args.remote_copy:    # Leave everything as is
```
Comment (clebergnu) [inline avocado/core/remote/runner.py]: OK.

File: avocado/core/remote/runner.py [hunk 4]
```diff
@@ -367,7 +367,7 @@ def setup(self):
         self.job.args.remote_username = self.job.args.vm_username
         self.job.args.remote_password = self.job.args.vm_password
         self.job.args.remote_key_file = self.job.args.vm_key_file
-        self.job.args.remote_no_copy = self.job.args.vm_no_copy
+        self.job.args.remote_copy = self.job.args.vm_copy
```
Comment (clebergnu) [inline avocado/core/remote/runner.py]: OK.

File: docs/source/RunningTestsRemotely.rst [hunk 5]
```diff
@@ -62,8 +62,10 @@ Once the remote machine is properly setup, you may run your test. Example::
     RESULTS    : PASS 1 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0
     TESTS TIME : 1.01 s
 
-As you can see, Avocado will copy the tests you have to your remote machine and
-execute them. A bit of extra logging information is added to your job summary,
+Notice that Avocado will not copy the tests to the remote machine. If you want
```
Comment (clebergnu) [inline docs/source/RunningTestsRemotely.rst]: I'll add either a section, or some more in-place information, about the caveats of Avocado's remote test copying support.

File: avocado/core/job.py (hunk 1 above)
Comment (apahim) [inline avocado/core/job.py]: Yes, it does make sense. Working on that.

File: avocado/core/remote/runner.py (hunk 2 above)
Comment (apahim) [inline avocado/core/remote/runner.py]: Sure, let me amend the commit message.

File: docs/source/RunningTestsRemotely.rst (hunk 5 above)
Comment (apahim) [inline docs/source/RunningTestsRemotely.rst]: Ok, I'll improve the docs. Thank you.

File: avocado/core/job.py (hunk 1 above)
Comment (apahim) [inline avocado/core/job.py]: Well, I giving this a thought now... I'm not sure the loader should be engaged in sort out when a test is remote or not. We have the job, which is local, the RemoteRunner, which runs the test remotely and the loader, which IMO should not be called by the local job when the test is remote. The job is the entity aware about the "remote" principle of the runner.
Anyway, I agree faking the suite is not the best approach. We should instead accept `None` as a valid value for suite and don't try to handle the suite when it's `None`. loader will then only be involved when the test is local. @clebergnu how does this sound for you?

File: avocado/plugins/remote.py [hunk 6]
```diff
@@ -68,10 +68,10 @@ def configure(self, parser):
...
                                         'a private key instead of a password '
                                         '(Example: .pem files from Amazon EC2)')
-        self.remote_parser.add_argument('--remote-no-copy',
-                                        dest='remote_no_copy',
+        self.remote_parser.add_argument('--remote-copy',
+                                        dest='remote_copy',
                                         action='store_true',
-                                        help="Don't copy tests and use the "
+                                        help="Copy tests and use the "
                                         "exact uri on guest machine.")
```
Comment (adereis) [inline avocado/plugins/remote.py]: Why this reference to "exact uri on guest machine" here?

File: avocado/plugins/vm.py [hunk 7]
```diff
@@ -74,8 +74,8 @@ def configure(self, parser):
                                     action='store_true', default=False,
                                     help='Restore VM to a previous state, '
                                     'before running tests')
-        self.vm_parser.add_argument('--vm-no-copy', action='store_true',
-                                    help="Don't copy tests and use the "
+        self.vm_parser.add_argument('--vm-copy', action='store_true',
+                                    help="Copy tests and use the "
                                     "exact uri on VM machine.")
```
Comment (adereis) [inline avocado/plugins/vm.py]: ditto

File: avocado/plugins/remote.py (hunk 6 above)
Comment (apahim) [inline avocado/plugins/remote.py]: Thinking about it, seems like it makes sense for the previous message, because we would look for the test in the exact path provided as url. But now with the inverted logics, it makes no sense to say that anymore.

File: avocado/plugins/vm.py (hunk 7 above)
Comment (apahim) [inline avocado/plugins/vm.py]: yep. fixing.

File: avocado/plugins/remote.py [hunk 8]
```diff
@@ -68,10 +68,10 @@ def configure(self, parser):
                                         help='Specify an identity file with '
                                         'a private key instead of a password '
                                         '(Example: .pem files from Amazon EC2)')
-        self.remote_parser.add_argument('--remote-no-copy',
-                                        dest='remote_no_copy',
+        self.remote_parser.add_argument('--remote-copy',
+                                        dest='remote_copy',
                                         action='store_true',
-                                        help="Don't copy tests and use the "
+                                        help="Copy tests and use the "
```
Comment (adereis) [inline avocado/plugins/remote.py]: I think all these `--copy` options should be documented as "Try to copy tests to...". There are many use-cases that will break with `--copy`, especially as we implement more advanced features, like multiplexing.

Or, like I would prefer, they should be removed, as they don't match the architecture of Avocado, where tests are not necessarily files. If removed, the documentation should provide an example of rsync invocation to copy tests, data-dirs, etc, which could be as simple as:

`$ cd avocado`
`$ rsync -a examples/tests -e ssh user@remote:~/avocado`
`$ avocado run --remote ...`

Or as a single line:

`$ rsync -a examples/tests -e ssh user@remote:~/avocado && avocado run --remote ...`

Comment (apahim) [inline avocado/plugins/remote.py]: OK, I was reluctant about removing them to not impact users that much, but since we are going to change the behaviour anyway, let's remove them once for all.

**Comment by adereis** [comment]
BTW, a few use-cases which I would expect should work after your changes if the two sides are perfectly in sync (avocado and all dependencies -- imagine a perfect scenario where the two machines sharing the same file-system):

 - run a remote test with the multiplexer
 - run avocado-vt tests remotely (`avocado run usb.usb_boot.usb_kbd.without_usb_hub --remote ...`)
 - when running multiple tests, the Test Names / Test IDs in the UI should now be consistent: they should be the same, no matter if the tests are being run remotely or locally (for example, test names should not be converted to full paths when running remotely if they're relative -- this was a problem with the previous implementation)
 - record the output of a test locally (`--output-check-record`) and then run the test remotely -- this will check the datadir usage
 - basically, everything should work: asset fetcher, multiplexer, timeouts, replay, output formats (json, xunit, tap), html report (saved locally), external-runner, wrappers...

If these things didn't work before, then don't feel obliged to fix them on this PR, but they should be documented as bugs in trello.
